### PR TITLE
Call Sentry if attaching highlight fails

### PR DESCRIPTION
**Pull request by PiotrKozlowski**
for: https://github.com/openstax/unified/issues/875

TODO:
- [x] Add tests

### CODE REVIEW

File: src/app/content/components/utils/attachHighlight.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import Highlighter, { Highlight, SerializedHighlight } from '@openstax/highlighter';
+import Sentry from '../../../../helpers/Sentry';
+
+const attachHighlight = (
+  highlight: Highlight | SerializedHighlight,
+  highlighter: Highlighter
+) => {
+  if (highlight instanceof SerializedHighlight) {
```
Comment (TomWoodward) [inline src/app/content/components/utils/attachHighlight.ts]: you don't have to recurse here, it does it internally

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/app/content/components/utils/attachHighlight.ts (hunk 1 above)
Comment (PiotrKozlowski) [reply src/app/content/components/utils/attachHighlight.ts]: @TomWoodward It does, but `SerializedHighlight` has no `isAttached` method. I could write this as: (is this better)?
```
  highlighter.highlight(highlight);
  const hl = highlight instanceof SerializedHighlight
    ? highlighter.getHighlight(highlight.id)
    : highlight;
  if (!hl || !hl.isAttached()) {
    Sentry.captureException(`Highlight with id: ${highlight.id} has not been attached.`);
  }
```

Comment (PiotrKozlowski) [reply src/app/content/components/utils/attachHighlight.ts]: I've updated the code, but it looks like some tests from `highlightManager.spec.tsx` are failing
Probably some mocks needs updates

Comment (TomWoodward) [reply src/app/content/components/utils/attachHighlight.ts]: src/app/content/components/Page/highlightManager.ts already calls highlighter.getHighlight(data.id); after it calls highlight, you could have your attachHighlight utility do that and return the highlight to save it the trouble

File: src/app/content/components/utils/attachHighlight.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+const attachHighlight = (
+  highlight: Highlight | SerializedHighlight,
+  highlighter: Highlighter
+) => {
+  highlighter.highlight(highlight);
+  const hl = highlight instanceof SerializedHighlight
+    ? highlighter.getHighlight(highlight.id)
+    : highlight;
+  if (!hl || !hl.isAttached()) {
+    Sentry.captureException(`Highlight with id: ${highlight.id} has not been attached.`);
```
Comment (TomWoodward) [inline src/app/content/components/utils/attachHighlight.ts]: captureException takes an `Error object`, i think you want `Sentry.error()` which takes a string

File: src/app/content/components/Page/highlightManager.ts [hunk 3]
```diff
@@ -87,7 +88,7 @@ const highlightData = (services: Services) => (data: HighlightData) => {
 
   const serialized = SerializedHighlight.fromApiResponse(data);
 
-  highlighter.highlight(serialized);
+  attachHighlight(serialized, highlighter);
 
   return highlighter.getHighlight(data.id);
```
Comment (TomWoodward) [inline src/app/content/components/Page/highlightManager.ts]: you could have `attachHighlight` return the highlight object so you don't have to fetch it a second time here

File: src/app/content/search/utils.ts [hunk 4]
```diff
@@ -149,7 +150,7 @@ export const highlightResults = (
     const hitHighlights = hit.highlight.visibleContent.map((highlightText, index) => {
       const highlights = getHighlightRanges(element, highlightText).map((range) => {
         const highlight = new Highlight(range.nativeRange, {content: range.toString()});
-        highlighter.highlight(highlight);
+        attachHighlight(highlight, highlighter);
```
Comment (TomWoodward) [inline src/app/content/search/utils.ts]: in the case of search highlights, these results will not have meaningful ids, so the error message is not really helpful

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/app/content/search/utils.ts (hunk 4 above)
Comment (PiotrKozlowski) [reply src/app/content/search/utils.ts]: What error message should I pass then?

Comment (TomWoodward) [reply src/app/content/search/utils.ts]: we want to be able to identify what result the problem happened with, but search results don't have a unique id. so i think the best we can do is include the search query, book, and page. maybe the index of the result in the results for the page too?

Comment (PiotrKozlowski) [reply src/app/content/search/utils.ts]: ok, let's do this in a separate card then

**Review by TomWoodward — APPROVED** [review]

**Pull request merged**